tests/csapi/thread_notifications_test.go

```diff
@@ -0,0 +1,178 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/runtime"
+)
+
+func syncHasUnreadNotifs(roomID string, check func(gjson.Result, gjson.Result) bool) client.SyncCheckOpt {
+	return func(clientUserID string, topLevelSyncJSON gjson.Result) error {
+		unreadNotifications := topLevelSyncJSON.Get("rooms.join." + client.GjsonEscape(roomID) + ".unread_notifications")
+		unreadThreadNotifications := topLevelSyncJSON.Get("rooms.join." + client.GjsonEscape(roomID) + ".unread_thread_notifications")
+		if !unreadNotifications.Exists() {
+			return fmt.Errorf("syncHasUnreadNotifs(%s): missing notifications", roomID)
+		}
+		if check(unreadNotifications, unreadThreadNotifications) {
+			return nil
+		}
+		return fmt.Errorf("syncHasUnreadNotifs(%s): check function did not pass: %v / %v", roomID, unreadNotifications.Raw, unreadThreadNotifications.Raw)
+	}
+}
+
+// Test behavior of threaded receipts and notifications.
+//
+// Send a series of messages, some of which are in threads. Send combinations of
+// threaded and unthreaded receipts and ensure the notification counts are updated
+// appropriately.
+func TestThreadedReceipts(t *testing.T) {
+	runtime.SkipIf(t, runtime.Dendrite) // not supported
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	// Create a room with alice and bob.
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	bob.JoinRoom(t, roomID, nil)
+
+	token := bob.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(bob.UserID, roomID))
+
+	// Send messages as alice and then check the highlight and notification counts from bob.
+	firstEventID := alice.SendEventSynced(t, roomID, b.Event{
```

Comment:
```suggestion
	firstHighlightEventForBobID := alice.SendEventSynced(t, roomID, b.Event{
```

Related to https://github.com/matrix-org/complement/pull/496#discussion_r996208469

Reply:
It might also help to refer to the names `A, B, C, D` from the diagram?